bin/e2hs-writer/src/reader.rs

```diff
@@ -0,0 +1,388 @@
+        let payload = block.body.execution_payload.clone();
+        let transactions = decode_transactions(&payload.transactions)?;
+        let header = pre_capella_execution_payload_to_header(payload.clone(), &transactions)?;
+        let historical_batch = HistoricalBatch {
+            state_roots: era.era_state.state.state_roots().clone(),
+            block_roots: era.era_state.state.block_roots().clone(),
+        };
+        let slot = block.slot;
+
+        // create beacon block proof
+        let historical_batch_proof = historical_batch.build_block_root_proof(slot % EPOCH_SIZE);
+        let beacon_block_proof: FixedVector<B256, typenum::U14> = historical_batch_proof.into();
+
+        // create execution block proof
+        let mut execution_block_hash_proof = block.body.build_execution_block_hash_proof();
+        let body_root_proof = block.build_body_root_proof();
+        execution_block_hash_proof.extend(body_root_proof);
+        let execution_block_proof: FixedVector<B256, typenum::U11> =
+            execution_block_hash_proof.into();
+
+        let proof = BlockProofHistoricalRoots {
+            beacon_block_proof,
+            beacon_block_root: block.tree_hash_root(),
+            slot,
+            execution_block_proof,
+        };
+
+        let header_with_proof = HeaderWithProof {
+            header,
+            proof: BlockHeaderProof::HistoricalRoots(proof),
+        };
+        let body = BlockBody(AlloyBlockBody {
+            transactions,
+            ommers: vec![],
+            withdrawals: None,
+        });
+        let receipts = self
+            .execution_api
+            .get_era_receipts(
+                block_number,
+                payload.transactions.len() as u64,
+                payload.receipts_root,
+            )
+            .await?;
+        Ok(AllBlockData {
+            block_number,
+            header_with_proof,
+            body,
+            receipts,
+        })
+    }
+
+    async fn get_pre_deneb_block_data(&self, block_number: u64) -> anyhow::Result<AllBlockData> {
+        let era = self.era_provider.get_era_for_block(block_number)?;
+        let block = era
+            .blocks
+            .iter()
+            .find(|block| block.block.execution_block_number() == block_number)
+            .ok_or_else(|| {
+                anyhow!("Era file for block #{block_number} not found during pre-deneb lookup")
+            })?;
+        let block = block
+            .block
+            .message_capella()
+            .map_err(|e| anyhow!("Unable to decode capella block: {e:?}"))?;
+        let payload = block.body.execution_payload.clone();
+        let transactions = decode_transactions(&payload.transactions)?;
+        let withdrawals: Vec<Withdrawal> =
+            payload.withdrawals.iter().map(Withdrawal::from).collect();
+        let header = pre_deneb_execution_payload_to_header(
+            payload.clone(),
+            &transactions,
+            withdrawals.clone(),
```

Comment:
```suggestion
        let withdrawals: Vec<Withdrawal> =
            payload.withdrawals.iter().map(Withdrawal::from).collect();
        let header = pre_deneb_execution_payload_to_header(
            payload.clone(),
            &transactions,
            &withdrawals,
```

I tested locally and we shouldn't need to clone withdrawals here